File: src/csharp/29 - Divide Two Integers.cs
class Solution
{
    // Time Limit Exceeded
    static int Divide(int dividend, int divisor)
    {
        switch (divisor)
        {
            case 1:
                return dividend;
            case -1 when dividend == int.MinValue:
                return int.MaxValue;
            case -1:
                return -dividend;
            default:
                break;
        }
        var result = 0;
        switch (dividend, divisor) 
        {
            case var _ when dividend > 0 && divisor < 0: 
                while (dividend >= 0)
                {
                    dividend += divisor;
                    result --;
                }
                result ++;
                break;
            case var _ when dividend < 0 && divisor > 0: 
                while (dividend <= 0)
                {
                    dividend += divisor;
                    result --;
                }
                result ++;
                break;
            case var _ when dividend < 0 && divisor < 0: 
                while (dividend <= divisor)
                {
                    dividend -= divisor;
                    result ++;
                }
                break;
            default:    // dividend >= 0 && divisor > 0
                while (dividend >= divisor)
                {
                    dividend -= divisor;
                    result ++;
                }
                break;
        }

        return result;
    }

    // ============================================

    // proper solution
    static int Divide(int dividend, int divisor)
    {
        static int InnerFunc(in int val, in int dividend, in int divisor)
        {
            if ((dividend > 0 ? -dividend : dividend) > (divisor > 0 ? -divisor : divisor))
            {
                return val;
            }
            var (k, v) = ((dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0)) ? (1, divisor) : (-1, -divisor);
            var pre = (k, v);
            while ((v > 0 && dividend > v && v <= (int.MaxValue >> 1)) || (v < 0 && dividend < v && v >= (int.MinValue >> 1)))
            {
                pre = (k, v);
                (k, v) = (k << 1, v << 1);
            }
            return InnerFunc(val + pre.k, dividend - pre.v, divisor);
        }

        // 结果溢出
        if (dividend == int.MinValue && divisor == -1)
        {
            return int.MaxValue;
        }
        else
        {
            return InnerFunc(0, dividend, divisor);
        }
    }

    static void Main(string[] args)
    {
        Console.WriteLine(Divide(int.MaxValue, 2));
        Console.WriteLine(Divide(1100540749, -1090366779));
        Console.WriteLine(Divide(int.MaxValue / 2, int.MinValue));
        Console.WriteLine(Divide(int.MinValue, -3));
        Console.WriteLine(Divide(int.MaxValue, 3));
        Console.WriteLine(Divide(int.MinValue, 2));
        Console.WriteLine(Divide(10, 3));
        Console.WriteLine(Divide(7, -3));
        Console.WriteLine(Divide(0, 1));
        Console.WriteLine(Divide(1, 1));
        Console.WriteLine(Divide(-10, -3));
        Console.WriteLine(Divide(int.MinValue, -1));
        Console.WriteLine(Divide(1, -1));
        Console.WriteLine(Divide(12, 3));
        Console.WriteLine(Divide(-12, 3));
        Console.WriteLine(Divide(12, -3));
        Console.WriteLine(Divide(-12, -3));
        Console.WriteLine(Divide(int.MinValue, 1));
    }
}

File: src/fsharp/29 - Divide Two Integers.fs
open System

let divide dividend divisor =
    // 快速判断是否到达溢出临界点
    let half0 = Int32.MaxValue >>> 1
    let half1 = Int32.MinValue >>> 1
    // 映射到负整数域、规避溢出问题
    let inline negative x = if x > 0 then -x else x
    // 判断是否同号
    let inline same x y = (x > 0 && y > 0) || (x < 0 && y < 0)
    // 倍增被除数、减法实现除法、尾递归实现
    let rec fn value dividend divisor =
        if negative dividend > negative divisor then value
        else
            let mutable (k, v) = if (same dividend divisor) then (1, divisor) else (-1, -divisor)
            let mutable (pk, pv) = (k, v)
            while (v > 0 && dividend > v && v <= half0) || (v < 0 && dividend < v && v >= half1) do
                pk <- k
                pv <- v
                k <- k <<< 1
                v <- v <<< 1
            fn (value + pk) (dividend - pv) divisor
    match (dividend, divisor) with
    | (_, 1) -> dividend
    | (Int32.MinValue, -1) -> Int32.MaxValue
    | (_, -1) -> -dividend
    | _ -> fn 0 dividend divisor

[<EntryPoint>]
let main argv =
    (Int32.MaxValue, 2) ||> divide |> printfn "%d"
    (1100540749, -1090366779) ||> divide |> printfn "%d"
    (Int32.MaxValue / 2, Int32.MinValue) ||> divide |> printfn "%d"
    (Int32.MinValue, -3) ||> divide |> printfn "%d"
    (Int32.MaxValue, 3) ||> divide |> printfn "%d"
    (Int32.MinValue, 2) ||> divide |> printfn "%d"
    (Int32.MinValue, -1) ||> divide |> printfn "%d"
    (Int32.MinValue, 1) ||> divide |> printfn "%d"
    (10, 3) ||> divide |> printfn "%d"
    (-10, -3) ||> divide |> printfn "%d"
    (7, -3) ||> divide |> printfn "%d"
    (0, 1) ||> divide |> printfn "%d"
    (1, 1) ||> divide |> printfn "%d"
    (12, 3) ||> divide |> printfn "%d"
    (-12, -3) ||> divide |> printfn "%d"
    (12, -3) ||> divide |> printfn "%d"
    (-12, 3) ||> divide |> printfn "%d"
    0